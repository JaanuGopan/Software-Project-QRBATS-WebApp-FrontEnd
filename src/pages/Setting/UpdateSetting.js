import React, { useState } from "react";
import "./Setting.css";
import eventCreationImage from "../../assets/Images/signin/Signin.jpeg";
import { IoMdCloseCircleOutline } from "react-icons/io";

const UpdateSetting = ({ handlecloseUpdateSettingWindow }) => {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [userName, setUserName] = useState("");
  const [password, setPassword] = useState("");

  return (
    <div className="setting-update-main-container">
      <div
        className="closeCreateEventWindow"
        onClick={handlecloseUpdateSettingWindow}
      >
        <IoMdCloseCircleOutline />
      </div>
      <h2>Update Profile</h2>
      <div className="setting-field">
        <img src={eventCreationImage} className="Create-logo" alt="Logo" />
        <div className="eventCreation-input-field">
          <form onSubmit={""}>
            <div className="Setting-input-with-icon">
              <label>Admin Name</label>
              <input
                type="text"
                id="eventName"
                name="eventName"
                placeholder={"Admin Name"}
                className="form-control mb-2"
                value={""}
                onChange={""}
              />
            </div>
            <div className="Setting-input-with-icon">
              <label>Username</label>
              <input
                type="text"
                id="eventName"
                name="eventName"
                placeholder={"Username"}
                className="form-control mb-2"
                value={""}
                onChange={""}
              />
            </div>
            <div className="Setting-input-with-icon">
              <label>Password</label>
              <input
                type="text"
                id="eventName"
                name="eventName"
                placeholder={"Password"}
                className="form-control mb-2"
                value={""}
                onChange={""}
              />
            </div>
            <button onClick={""} className="btn btn-primary w-100">
              Save
            </button>
          </form>
        </div>
      </div>
    </div>
  );
};

export default UpdateSetting;
