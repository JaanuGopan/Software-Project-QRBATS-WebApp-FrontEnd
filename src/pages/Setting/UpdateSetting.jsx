import Switch from '@mui/material/Switch';
import axios from 'axios';
import React, { useContext, useEffect, useState } from 'react';
import toast, { Toaster } from 'react-hot-toast';
import { IoMdCloseCircleOutline } from 'react-icons/io';
import { useDispatch } from 'react-redux';
import Select from 'react-select';
import UserService from '../../api/services/UserService';
import InputField from '../../components/textfields/InputBox/InputField';
import InputPassword from '../../components/textfields/InputPassword/InputPassword';
import WarningPopup from '../../components/warningPopup/WarningPopup';
import { AuthContext } from '../../config/AuthProvider';
import { resetSideBarIndex } from '../../redux/features/mainNavigationSlice';
import Department from '../../utils/Department';
import './Setting.css';

const UpdateSetting = ({ handleCloseUpdateSettingWindow }) => {
  const dispatch = useDispatch();
  const { user } = useContext(AuthContext);
  const { logout } = useContext(AuthContext);

  const [firstName, setFirstName] = useState('');
  const [lastName, setLastName] = useState('');
  const [email, setEmail] = useState('');
  const [userRole, setUserRole] = useState();
  const [departmentId, setDepartmentId] = useState({});
  const [userName, setUserName] = useState('');
  const [oldPassword, setOldPassword] = useState('');
  const [newPassword, setNewPassword] = useState('');
  const [confirmNewPassword, setConfirmNewPassword] = useState('');
  const [errors, setErrors] = useState({});
  const [showOldPassword, setShowOldPassword] = useState(false);
  const [showNewPassword, setShowNewPassword] = useState(false);
  const [isLoading, setIsLoading] = useState(false);

  const departmentList = Department.departmentList.map((value, index) => ({
    value: index + 1,
    label: value,
  }));

  const notifySuccess = () => toast.success('Successfully Profile Updated!');
  const notifyPasswordVerify = () => toast.success('Successfully Password Verified!');

  useEffect(() => {
    if (user) {
      const response = axios.get('/api/v1/auth/user').then((res) => {
        setUserName(res.data.userName);
        setFirstName(res.data.firstName);
        setLastName(res.data.lastName);
        setEmail(res.data.email);
        setUserRole(res.data.role);
        setDepartmentId({
          value: res.data.departmentId,
          label: Department.departmentList[res.data.departmentId - 1],
        });
      });
    }
  }, []);

  const handleShowOldPassword = (event) => {
    setShowOldPassword(event.target.checked);
    setShowNewPassword(false);
    setOldPassword('');
  };

  const validateEmail = (email) => {
    const re = /\S+@\S+\.\S+/;
    return re.test(email);
  };

  const handleValidation = () => {
    let formErrors = {};
    let formIsValid = true;

    if (!firstName) {
      formIsValid = false;
      formErrors['firstName'] = 'First Name cannot be empty';
    }

    if (!lastName) {
      formIsValid = false;
      formErrors['lastName'] = 'Last Name cannot be empty';
    }

    if (!email) {
      formIsValid = false;
      formErrors['email'] = 'Email cannot be empty';
    } else if (!validateEmail(email)) {
      formIsValid = false;
      formErrors['email'] = 'Email is not valid';
    }

    if (!userName) {
      formIsValid = false;
      formErrors['userName'] = 'User Name cannot be empty';
    }

    if (showNewPassword) {
      if (!newPassword) {
        formIsValid = false;
        formErrors['newPassword'] = 'New Password cannot be empty';
      }

      if (newPassword !== confirmNewPassword) {
        formIsValid = false;
        formErrors['confirmNewPassword'] = 'Passwords do not match';
      }
    }

    setErrors(formErrors);
    return formIsValid;
  };

  const handleLogoutClick = () => {
    dispatch(resetSideBarIndex());
    logout();
  };

  const [loadingUpdateUser, setLoadingUpdateUser] = useState(false);
  const handleUpdateUser = async (event) => {
    event.preventDefault();
    if (handleValidation()) {
      try {
        setLoadingUpdateUser(true);
        const response = await UserService.updateUser(
          user.userId,
          firstName,
          lastName,
          email,
          userName,
          newPassword,
          departmentId.value
        );
        if (response.status === 200) {
          handleCloseUpdateSettingWindow();
          notifySuccess();
          handleLogoutClick();
        } else if (response.status === 400) {
          setErrors(response.data);
          toast.error(response.data);
          setShowUpdateWarning(false);
        }
      } finally {
        setLoadingUpdateUser(false);
      }
    } else {
      setShowUpdateWarning(false);
    }
  };

  const handleVerifyPassword = async (e) => {
    e.preventDefault();
    if (oldPassword) {
      setIsLoading(true);
      try {
        const response = await UserService.verifyPassword(userName, oldPassword);
        if (response === true) {
          setShowNewPassword(true);
          notifyPasswordVerify();
        }
      } catch (error) {
        console.error(error);
      } finally {
        setIsLoading(false);
      }
    }
  };

  const [showUpdateWarning, setShowUpdateWarning] = useState(false);

  const handleShowUpdateWarning = (e) => {
    e.preventDefault();
    setShowUpdateWarning(true);
  };

  return (
    <div className="setting-update-main-container">
      <Toaster />
      <div className="setting-update-title-close-button">
        <h3 className="setting-update-title">Update Profile</h3>
        <div className="setting-update-close-button" onClick={handleCloseUpdateSettingWindow}>
          <IoMdCloseCircleOutline id="close-icon" />
        </div>
      </div>
      <div className="setting-field">
        <div className="update-setting-input-form">
          <form onSubmit={handleShowUpdateWarning}>
            <div className="Setting-input-with-icon">
              <label>First Name</label>
              <InputField
                required
                inputType={'text'}
                placeholder={'Enter First Name'}
                value={firstName}
                onChange={(e) => setFirstName(e.target.value)}
                error={!!errors.firstName}
                helperText={errors.firstName}
              />
            </div>
            <div className="Setting-input-with-icon">
              <label>Last Name</label>
              <InputField
                inputType={'text'}
                placeholder={'Enter Last Name'}
                value={lastName}
                onChange={(e) => setLastName(e.target.value)}
                error={!!errors.lastName}
                helperText={errors.lastName}
              />
            </div>
            <div className="Setting-input-with-icon">
              <label>Email</label>
              <InputField
                inputType={'text'}
                placeholder={'Enter Email'}
                value={email}
                onChange={(e) => setEmail(e.target.value)}
                error={!!errors.email}
                helperText={errors.email}
              />
            </div>
            {user.role === 'LECTURER' && (
              <div className="Setting-input-with-icon">
                <label>Department</label>
                <Select
                  required
                  id="department"
                  name="department"
                  placeholder={'Select Department'}
                  options={departmentList}
                  value={departmentId}
                  onChange={(e) => setDepartmentId(e)}
                />
              </div>
            )}
            <div className="Setting-input-with-icon">
              <label>User Name</label>
              <InputField
                inputType={'text'}
                placeholder={'Enter UserName'}
                value={userName}
                onChange={(e) => setUserName(e.target.value)}
                error={!!errors.userName}
                helperText={errors.userName}
              />
            </div>
            <div>
              <label>Change Password</label>
              <Switch
                checked={showOldPassword}
                onChange={handleShowOldPassword}
                inputProps={{ 'aria-label': 'controlled' }}
              />
            </div>

            {showOldPassword === true && showNewPassword === false && (
              <div className="Setting-input-with-icon">
                <label>Old Password</label>
                <div className="col">
                  <InputPassword
                    placeholder={'Enter Old Password'}
                    value={oldPassword}
                    onChange={(e) => setOldPassword(e.target.value)}
                    buttonText={'verify'}
                    onButtonClick={(e) => handleVerifyPassword(e)}
                    isLoading={isLoading}
                    error={!!errors.oldPassword}
                    helperText={errors.oldPassword}
                  />
                </div>
              </div>
            )}
            {showOldPassword && showNewPassword && (
              <div className="Setting-input-with-icon">
                <label>New Password</label>
                <InputPassword
                  placeholder={'Enter new Password'}
                  value={newPassword}
                  onChange={(e) => setNewPassword(e.target.value)}
                  showButton={false}
                  error={!!errors.newPassword}
                  helperText={errors.newPassword}
                />
                <label>Confirm New Password</label>
                <InputPassword
                  placeholder={'Renter new Password'}
                  value={confirmNewPassword}
                  onChange={(e) => setConfirmNewPassword(e.target.value)}
                  showButton={false}
                  error={!!errors.confirmNewPassword}
                  helperText={errors.confirmNewPassword}
                />
              </div>
            )}

            <div className="settings-update-button">
              <button type="submit" className="btn btn-warning w-100 mt-3">
                Save
              </button>
            </div>
          </form>
        </div>
      </div>
      {showUpdateWarning && (
        <div className="setting-update-warning-container">
          <WarningPopup
            handleOk={handleUpdateUser}
            titleText={'If You Want To Update, You Need To Login Again'}
            buttonText={'Update'}
            handleCloseWarningWindow={() => setShowUpdateWarning(false)}
            processing={loadingUpdateUser}
          />
        </div>
      )}
    </div>
  );
};

export default UpdateSetting;
