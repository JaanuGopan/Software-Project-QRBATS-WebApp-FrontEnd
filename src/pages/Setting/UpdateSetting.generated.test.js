import renderer from 'react-test-renderer';
import React, { useState } from "react";

import eventCreationImage from "../../assets/Images/signin/Signin.jpeg";
import { IoMdCloseCircleOutline } from "react-icons/io";
import UpdateSetting from "./UpdateSetting";

jest.mock("./Setting.css");
jest.mock("../../assets/Images/signin/Signin.jpeg");
jest.mock("react-icons/io");

const renderTree = tree => renderer.create(tree);
describe('<UpdateSetting>', () => {
  it('should render component', () => {
    expect(renderTree(<UpdateSetting 
    />).toJSON()).toMatchSnapshot();
  });
  
});