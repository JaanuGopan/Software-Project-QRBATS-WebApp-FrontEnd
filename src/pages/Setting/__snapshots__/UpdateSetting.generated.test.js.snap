// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpdateSetting> should render component 1`] = `
<div
  className="setting-update-main-container"
>
  <div
    className="closeCreateEventWindow"
  />
  <h2>
    Update Profile
  </h2>
  <div
    className="setting-field"
  >
    <img
      alt="Logo"
      className="Create-logo"
      src="Signin.jpeg"
    />
    <div
      className="eventCreation-input-field"
    >
      <form
        onSubmit=""
      >
        <div
          className="Setting-input-with-icon"
        >
          <label>
            Admin Name
          </label>
          <input
            className="form-control mb-2"
            id="eventName"
            name="eventName"
            onChange=""
            placeholder="Admin Name"
            type="text"
            value=""
          />
        </div>
        <div
          className="Setting-input-with-icon"
        >
          <label>
            Username
          </label>
          <input
            className="form-control mb-2"
            id="eventName"
            name="eventName"
            onChange=""
            placeholder="Username"
            type="text"
            value=""
          />
        </div>
        <div
          className="Setting-input-with-icon"
        >
          <label>
            Password
          </label>
          <input
            className="form-control mb-2"
            id="eventName"
            name="eventName"
            onChange=""
            placeholder="Password"
            type="text"
            value=""
          />
        </div>
        <button
          className="btn btn-primary w-100"
          onClick=""
        >
          Save
        </button>
      </form>
    </div>
  </div>
</div>
`;
